[ Sequoia wk2 ] two accessibility/mac/ pdf tests are a consistent timeout

hoffmanjoshua opened the pull request:
#### 85e2ede7862a7216edda0e6eac8e29e97ec01f78
<pre>
[ Sequoia wk2 ] two accessibility/mac/ pdf tests are a consistent timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=281531">https://bugs.webkit.org/show_bug.cgi?id=281531</a>
<a href="https://rdar.apple.com/137996591">rdar://137996591</a>

Reviewed by Tyler Wilcock.

These PDF tests assumed that the accessibility hierarchy that is vended from PDFKit
does not change. But, between Sonoma and Sequoia, the structure had changed, causing
these tests with pre-defined paths to fail (specifically, when finding the AXPage and
AXStaticText elements).

These are now updated to use a &quot;search&quot; approach to find elements, rather than
a specific path, to avoid hierarchy assumptions and to increase their resilliency
against future changes.

* LayoutTests/accessibility/mac/basic-embed-pdf-accessibility-expected.txt:
* LayoutTests/accessibility/mac/basic-embed-pdf-accessibility.html:
* LayoutTests/accessibility/mac/iframe-pdf-expected.txt:
* LayoutTests/accessibility/mac/iframe-pdf.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:

Canonical link: <a href="https://commits.webkit.org/288850@main">https://commits.webkit.org/288850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6a0aeb7874ec325d98caa94b23e093d602d8aae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38985 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89736 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35648 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65855 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23670 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3343 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76882 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46115 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3222 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31094 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34723 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31893 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91110 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11936 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8703 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74316 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12163 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72686 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73438 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18163 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16246 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3375 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11888 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11722 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15216 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -371,3 +371,31 @@ function resetActiveElementAndSelectedChildren(id) {
+function findFirstPDFPage(startObject) {
+    if (!startObject) {
+        return null;
+    }
```
Do we need these curly braces?

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -371,3 +371,31 @@ function resetActiveElementAndSelectedChildren(id) {
+    let currentObject = startObject;
+    while (currentObject && currentObject.role != "AXRole: AXPage") {
+        currentObject = currentObject.childAtIndex(0);
+    }
```
Would it be more resilient to have a proper BFS / DFS search, rather than only checking the first child of everything on the way down? As one theoretical example — what if PDFKit decided to put a scrollbar as the first child, and pages as the following ones? Not saying that's likely...but it's not impossible either 🤔 

Do we need these curly braces?

hoffmanjoshua replied:
Yeah, this should be traversing all of the children. I can make this a true DFS

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -371,3 +371,31 @@ function resetActiveElementAndSelectedChildren(id) {
 }
+
+// PDF Test Helpers
+function findFirstPDFPage(startObject) {
```
There's nothing PDF-specific about this function, is there? Maybe we could call it `findFirstPageDescendant` or similar?

hoffmanjoshua replied:
That's a good point—I can change this name

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -371,3 +371,31 @@ function resetActiveElementAndSelectedChildren(id) {
+function traverseChildrenToFirstStaticText(startObject) {
+    if (!startObject || startObject.role == "AXRole: AXStaticText") {
+        return startObject;
+    }
```
Are these curly braces necessary?

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -371,3 +371,31 @@ function resetActiveElementAndSelectedChildren(id) {
+        let result = traverseChildrenToFirstStaticText(startObject.childAtIndex(i));
+        if (result) {
+            return result;
+        }
```
Are these curly braces necessary?

hoffmanjoshua committed:
[ Sequoia wk2 ] two accessibility/mac/ pdf tests are a consistent timeout
https://bugs.webkit.org/show_bug.cgi?id=281531
rdar://137996591

Reviewed by Tyler Wilcock.

These PDF tests assumed that the accessibility hierarchy that is vended from PDFKit
does not change. But, between Sonoma and Sequoia, the structure had changed, causing
these tests with pre-defined paths to fail (specifically, when finding the AXPage and
AXStaticText elements).

These are now updated to use a "search" approach to find elements, rather than
a specific path, to avoid hierarchy assumptions and to increase their resilliency
against future changes.

* LayoutTests/accessibility/mac/basic-embed-pdf-accessibility-expected.txt:
* LayoutTests/accessibility/mac/basic-embed-pdf-accessibility.html:
* LayoutTests/accessibility/mac/iframe-pdf-expected.txt:
* LayoutTests/accessibility/mac/iframe-pdf.html:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:

Canonical link: https://commits.webkit.org/288850@main
